### PR TITLE
Add support for qemu commandline in nova

### PR DESCRIPTION
**Pull request by wmousa**

### CODE REVIEW

File: nova/tests/unit/virt/libvirt/test_config.py [hunk 1]
```diff
@@ -1802,6 +1802,20 @@ def test_config_qga_full(self):
             </channel>""" % obj.source_path)
 
 
+class LibvirtConfigQemuCommandLineTest(LibvirtConfigBaseTest):
```
Comment (moshe010) [inline nova/tests/unit/virt/libvirt/test_config.py:1805]: can you test for multi-args?

File: nova/virt/libvirt/config.py [hunk 2]
```diff
@@ -2767,6 +2792,13 @@ def _format_devices(self, root):
             devices.append(dev.format_dom())
         root.append(devices)
 
+    def _format_qemu_args(self, root):
+        if len(self.qemu_args) == 0:
+            return
+        for qemu_arg in self.qemu_args:
```
Comment (moshe010) [inline nova/virt/libvirt/config.py:2796]: why do you need the Line 2797-2798. if the list is empty you won't enter it

File: nova/virt/libvirt/driver.py [hunk 3]
```diff
@@ -6009,12 +6016,30 @@ def _get_guest_config(self, instance, network_info, image_meta,
...
                 flavor, virt_type, self._host)
             guest.add_device(config)
 
+            if vif.get("vnic_type") == "virtio-forwarder":
+                # Enable page-per-vq for virtio-forwarder interface
+                args_dict = {"page-per-vq": "on"}
+                if (vif.get("network") and vif.get("network").get("meta") and
+                        vif.get("network").get("meta").get("mtu")):
+                    # Set mtu_host in xml file for virtio-forwarder interface
+                    args_dict["host_mtu"] = vif["network"]["meta"]["mtu"]
```
Comment (moshe010) [inline nova/virt/libvirt/driver.py:6034]: I think the host_mtu is a WA. Please add a note that we need to revisit if it need or not?

File: nova/virt/libvirt/driver.py [hunk 4]
```diff
@@ -6009,12 +6016,30 @@ def _get_guest_config(self, instance, network_info, image_meta,
...
+                qemu_args = self._get_guest_qemu_commandline("net%d"
+                                                             % net_index,
+                                                             **args_dict)
+                guest.qemu_args.append(qemu_args)
+            if vif.get("vnic_type") in ["virtio-forwarder", "normal", "macvtap"]:
```
Comment (moshe010) [inline nova/virt/libvirt/driver.py:6039]: can you check if we can set the alias name? then we can do vdpa0/ vdpa1 which will be less risky code

**Review by moshe010 — CHANGES_REQUESTED** [review]

File: nova/tests/unit/virt/libvirt/test_config.py (hunk 1 above)
Comment (wmousa) [reply nova/tests/unit/virt/libvirt/test_config.py:1805]: done

File: nova/virt/libvirt/config.py (hunk 2 above)
Comment (wmousa) [reply nova/virt/libvirt/config.py:2796]: done

File: nova/virt/libvirt/driver.py (hunk 3 above)
Comment (wmousa) [reply nova/virt/libvirt/driver.py:6034]: done

File: nova/virt/libvirt/driver.py (hunk 4 above)
Comment (wmousa) [reply nova/virt/libvirt/driver.py:6039]: I tried it before it's neglected, the nova code will export xml file without any alias names, I think the qemu is doing it

**Comment by moshe010** [comment]
@wmousa  you didn't push new commit

**Comment by wmousa** [comment]
@moshe010 The code is working fine and tested on a fresh deployment,
As this get merged, I'll open a PR on containerized-ovs project

**Review by moshe010 — APPROVED** [review]

**Comment by moshe010** [comment]
@wmousa  I don't have permission to merged but LGTM

**Pull request merged**